docs/proposals/20200804-windows-support.md

```diff
@@ -0,0 +1,271 @@
+CAPI and only Kubeadm bootstrap tokens, which have limited lifetime, should be used for the joining the Windows nodes. 
+
+### Risks and Mitigations
+
+- Privileged containers are not implemented.
+  - There is an active discussion and [KEP](https://docs.google.com/document/d/12EUtMdWFxhTCfFrqhlBGWV70MkZZPOgxw0X-LTR0VAo/edit#) In place.  At the Beta stage the community can do a checkpoint to determine if the solution fits customer needs
+- Cloudbase-init is a third party dependency
+  - This project is under Apache 2.0 License : https://github.com/cloudbase/cloudbase-init which is cleared under the CNCF Allow list: https://github.com/cncf/foundation/blob/master/allowed-third-party-license-policy.md
+- Windows VHD Distribution
+  - Cluster API already provides the ability to use customer provided VHD and images provided are only for testing demonstration purposes and it is recommended the customer creates their own image. 
+  - Customers using the VHD scripts must ensure they are following [Windows licensing requirements](https://www.microsoft.com/en-us/licensing/product-licensing/windows-server)
+- Wins is a third party dependency
+  - The project is under the Apache 2.0 License
+
+## Alternatives
+
+1. An alternative to using wins and daemonsets to do the configuration to download and configure components as services 
+   using the kubeadm pre/post commands. This would require the infrastructure providers to have the ability to pass 
+   configuration through the use of these commands which is already done today. During the Alpha phase with the pre/post 
+   scripts being developed by individual infra providers this will not be an issue.  With the move to Windows privileged
+   containers in Beta, this becomes a non issue as wins will no longer be required.
+
+1. Create a separate bootstrap provider for Windows.  This would require re-implementing a lot of the logic that is 
+   already in CABPK.  
+   When bugs are fixed or changes in behavior occur, Windows would risk being out of sync with the Linux implementation.
+
+1. Modified CABPK provider to have a different output format than cloud-init for Windows nodes.   With Cloudbase-init 
+   there are no requirements to change CABPK which makes the adaption for Windows straightforward. If we were to adapt 
+   the output format of CABPK there is a potential for introduction of bugs and variation in the logic that would be 
+   created for Windows nodes.  This would cause the Windows implementation to differ from others which could lead to 
+   confusion when debugging differences.
+
+## Upgrade Strategy
+
+Nodes that use this pattern will require the infrastructure to be immutable as specified by CAPI documentation.
+
+## Additional Details
+
+### Test Plan [optional]
+
+**Note:** *Section not required until targeted at a release.*
+
+Given there are no changes to the CAPI codebase, The testing plan would be left up to each infrastructure provider. 
```

Comment:
"Given there are no changes to the CAPI codebase" that's not entirely true since changes are required for bootstrap kubeadm provider which is part of the CAPI codebase. How are we planning on testing that? I'm especially curious about testplans for CAPI e2e using the docker provider.

Reply:
at the time of writing, there were no planned changes to the bootstrap provider. I will update with the new learning and a plan for testing the windows retry logic.  

Does the docker provider use kind to create the workload cluster?

Reply:
Looked into this a little further. 

The changes that would be proposed would be to add `UseExperimentalRetryJoin` for windows deployments.  There were concerns about this feature discussed in https://github.com/kubernetes-sigs/cluster-api/pull/2763#discussion_r397306055. I don't see that tests were added for this particular feature.  

It looks like the docker provider uses [kind](https://kind.sigs.k8s.io/).  Windows containers do not support docker in docker so we would not be able to reuse those tests even in the case that there were tests. I think we could add unit tests to ensure the correct script is output. 

It looks from the https://github.com/kubernetes-sigs/cluster-api/issues/2769#issuecomment-603989176 that there will be a need for retries in some fashion in CABPK.  I didn't really get a sense of the future direction for retries in cabpk, has there been further discussion on the future of `UseExperimentalRetryJoin`?

Reply:
I might start reviving the idea of a node agent that imports kubeadm as a library, which is a thing tat is happening and may call a meeting next week if people are interested. I think in the meantime, you could use a windows specific UseExperimentalRetryJoin. In which case cabpk would need to know if it's Linux or Windows the script is being generated for.

Reply:
> In which case cabpk would need to know if it's Linux or Windows the script is being generated for.

Spoke briefly with @CecileRobertMichon about where the best place for that to land OS information and came up with a few options.  It sounds like OS info might be needed for the [sentinel file in the bootstrap failure proposal](https://docs.google.com/document/d/1U0GxvO6ltgIINMjpQz96UD4bExlN2h21wyyn-3ENezc/edit) as well.

1. Add an additional OS field on the kubeadm spec next to [UseExperimentalRetryJoin](https://github.com/jsturtevant/cluster-api/blob/28b1087a10015beb421b51cbd4eade9653744710/bootstrap/kubeadm/api/v1alpha3/kubeadmbootstrapconfig_types.go#L97)
2. Add as Part of the contract for Machine spec and get it the same [k8s version is acquired](https://github.com/kubernetes-sigs/cluster-api/blob/fffcfedb09dab668b45b71c748f6609a53baa518/bootstrap/kubeadm/controllers/kubeadmconfig_controller.go#L806).

Option 2, feel more appropriate to me as the OS is a machine level construct and it could be reused across providers and other machinery. For Backwards compatibility and providers that wouldn't set it, it could default to Linux.

Reply:
2 seems fine to me since cabpk already introspects the Machine object today.

Reply:
I agree with @naadir that it seems right to put it on the Machine, but also that I think ideally we'll not need OS-specific kubeadm retry behavior. Inevitably there will be other things that need to switch on OS which is why it still makes sense to put it on the Machine so that it's not a one-off fix just for this problem.

Reply:
> ideally we'll not need OS-specific kubeadm retry behavior

would this be resolved via the "kubeadm as a library" idea?

Reply:
Just some clarifications

> kubeadm as a library, which is a thing tat is happening

This is being discussed with the kubeadm community, although I wouldn't expect anything to materialize before 2021.

> Spoke briefly with @CecileRobertMichon about where the best place for that to land OS information and came up with a few options. It sounds like OS info might be needed for the sentinel file in the bootstrap failure proposal as well.

We were talking for a different effort (https://github.com/kubernetes-sigs/cluster-api/pull/3668#issuecomment-696211676) to add a condition to the InfrastructureMachine that communicates "ready to receive bootstrap data".

The reason I'm bringing this up is that if in v1alpha4 we go down that route, we could have the infrastructure machine to optionally expose the operating system information based on the image provided. This information could be useful also for debugging purposes, or to write other code that's OS-dependent. While I just said "OS", we could have a well-defined struct with generic information that we want to expose.

Reply:
> would this be resolved via the "kubeadm as a library" idea?

Potentially. There's a whole bunch of use cases not just related to Windows. Will start jotting down together and convene a WG for working parties. This shouldn't block your initial implementation though.

Reply:
@vincepri is there a place to add for a note for an OS/generic struct to the infra machine for v1alpha4?

I believe it would it be adding a struct/field to: https://github.com/kubernetes-sigs/cluster-api/blob/b0236f2b3c1d0ee9e6a8dffd1baaa22fd8aee17d/api/v1alpha3/machine_types.go#L57

Reply:
IMO we should add the OS info to the Infra machine (AWSMachine, AzureMachine, DockerMachine, etc) since it's tightly coupled with the infrastructure and the OS image. Maybe it should be part of the infra provider contract?

Then we sync the InfraMachine with a Status field in the core CAPI Machine type, and use that in CABPK and any other bootstrap providers that want to support Windows.

Reply:
Yes that sounds good to me as well, I was hoping we'd get there. CAPI's Machine object should be agnostic and look as much as possible as a Kubernetes node, rather than a VM, which is the role of the infrastructure side of things

Reply:
I've add some wording to capture this in a distinct commit: https://github.com/kubernetes-sigs/cluster-api/pull/3616/commits/6ac0cb2a99930555d3e4908a39501aa7d1090aaa PTAL

Reply:
@rsmitty FYI since this would involve some changes around Bootstrap/Infra provider contract

Reply:
Seems fine to me. Adding OS info to the MetalMachine for our infra provider shouldn't be a big deal at all.

Reply:
I've updated this and added contract information in the docs PTAL

Reply:
Based on the conversation today at cluster-api meeting, the desire is remove experimental retry for for v1Alpha4. Since Windows support will be in v1alpha 3 I am going to drop the retry aspect. We can revisit the OsType field if a concrete use case comes up later.